### PR TITLE
Sort branches and tags by date descending

### LINKED ISSUE #5709: Branch sorting (branches and PR page)
- Gitea version (or commit ref): 176a604
- Git version: 2.17.1
- Operating system: ubuntu
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ X] Yes (provide example URL)
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description

Can branches be sorted by last commit date or maybe by date created? They are now sorted alphabetically and this doesn't help me much when I'd like to find my most recent branch.

I have to go and type it's name, it would be much easier to pick it if it was up front in list.

This goes for "Branches" page or when creating a new pull request.

## Screenshots

branches

Here, `test4` branch would come first in list if they were sorted by last commit date.

### LINKED ISSUE #17316: Sort tag listing by alphanumeric values
### Gitea Version

1.14.1

### Operating System

linux

### Browser Version

not relevant

### Can you reproduce the bug on the Gitea demo site?

Yes

### Description

For example on https://try.gitea.io/camlafit/ps_checkout# if we list tags on interface, listing doesn't follow semver list.
2.15.x  is after 2.1.x but before 2.2.x but 15 > 2.

Thanks a lot

### Screenshots

Capture d’écran de 2021-10-15 16-44-37

### PR DESCRIPTION
**Pull request by sergemedvid**
This PR fixes #5709 and #17316 by changing the order of listed branches and tags to show the ones with latest commits atop.
It's achieved with changing underlying "show-ref" git command with "for-each-ref" as suggested in https://stackoverflow.com/a/5188364
Also, it's passing format string so the output matches "show-ref" command output.

close #5709
close #17316

### CODE REVIEW

**Comment by silverwind** [comment]
This doesn't raise the minimum git version (currently 2.0.0), right?

**Comment by delvh** [comment]
According to the linked stackoverflow, it would raise it to `2.7.0`.

**Comment by sergemedvid** [comment]
> This doesn't raise the minimum git version (currently 2.0.0), right?

No, it doesn’t. The “for-each-ref” command was available in 2.0.0: https://github.com/git/git/blob/v2.0.0/Documentation/git-for-each-ref.txt

> According to the linked stackoverflow, it would raise it to `2.7.0`.

The linked stackoverflow mentions 2.7.0 in context that sorting became available for “git branch” command.

**Comment by delvh** [comment]
Yep, I misread that.

**Comment by silverwind** [comment]
`for-ref-each` seems to have been introduced in https://github.com/git/git/commit/9f613ddd21cbd05bfc139d9b1551b5780aa171f6, version 1.4.4.

File: modules/git/repo_branch_nogogit.go [hunk 1]
```diff
@@ -63,7 +63,7 @@ func (repo *Repository) IsBranchExist(name string) bool {
 // GetBranchNames returns branches from the repository, skipping skip initial branches and
 // returning at most limit branches, or all branches if limit is 0.
 func (repo *Repository) GetBranchNames(skip, limit int) ([]string, int, error) {
-	return callShowRef(repo.Ctx, repo.Path, BranchPrefix, "--heads", skip, limit)
+	return callShowRef(repo.Ctx, repo.Path, BranchPrefix, "refs/heads --sort=-committerdate", skip, limit)
```
Comment (zeripath) [inline modules/git/repo_branch_nogogit.go]: Can't we use:

```suggestion
	return callShowRef(repo.Ctx, repo.Path, BranchPrefix, BranchPrefix+" --sort=-committerdate", skip, limit)
```

and similarly elsewhere?

**Review by delvh — APPROVED** [review]

File: modules/git/repo_branch_nogogit.go (hunk 1 above)
Comment (sergemedvid) [reply modules/git/repo_branch_nogogit.go]: Good point, thank you! I've added a new commit which resolves it.

**Review by 6543 — APPROVED** [review]

File: modules/git/repo_branch_nogogit.go [hunk 2]
```diff
@@ -107,9 +107,9 @@ func walkShowRef(ctx context.Context, repoPath, arg string, skip, limit int, wal
 
 	go func() {
 		stderrBuilder := &strings.Builder{}
-		args := []string{"show-ref"}
+		args := []string{"for-each-ref", "--format=%(objectname) %(refname)"}
 		if arg != "" {
-			args = append(args, arg)
+			args = append(args, strings.Fields(arg)...)
```
Comment (wxiaoguang) [inline modules/git/repo_branch_nogogit.go:112]: Splitting arguments by space is dangerous.
* https://github.com/go-gitea/gitea/pull/21168

**Comment by iranvir** [comment]
Can someone please show how to turn this feature on in the Gitea UI? Is there a configuration line that needs to be added?

**Comment by 6543** [comment]
@iranvir if you dont use current dev version it's not released

**Pull request merged**